fix(compiler): unable to resolve destructuring variable declarations

crisbeto opened the pull request:
Currently the partial evaluator isn't able to resolve a variable declaration that uses destructuring in the form of `const {value} = {value: 0}; const foo = value;`. These changes add some logic to allow for us to resolve the variable's value.

Fixes #36917.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/partial_evaluator/test/evaluator_spec.ts:
```diff
@@ -200,6 +200,36 @@ runInEachFileSystem(() => {
       expect(evaluate('const a = null;', 'a')).toEqual(null);
     });
 
+    it('destructuring array variable declaration works', () => {
```
NIT: avoid "... works" descriptions.
I know that there are already some cases of this in this file but our general convention is to have a sentence that reads well with the `it` at the beginning... e.g. `it supports...` or `it should ...`.

```suggestion
    it('supports destructuring array variable declarations', () => {
```

petebacondarwin reviewed (approved):
LGTM but how far do we want to take this? For example more complex object binding patterns are not yet supported:

```ts
const { a: { b } } = { a: { b: 10 };
eval(b)
```

To support this we would need to traverse back up the AST tree from the `ts.BindingElement` tracking the `ts.ObjectBindingPattern` parents until we got to the `ts.VariableDeclaration` and then traverse back down the `initializer` using the `accessHelper()` to follow the correct properties/indexes.

crisbeto commented:
Good point, we should definitely handle the nested expressions. I was wondering what the use case would be where the parent is another `BindingElement`.

crisbeto commented:
Updated to handle nested expressions.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/partial_evaluator/test/evaluator_spec.ts:
```diff
@@ -200,6 +200,44 @@ runInEachFileSystem(() => {
+
+    it('supports nested destructuring array variable declarations', () => {
+      expect(evaluate(`const [[[a]]] = [[[1]]];`, 'a')).toBe(1);
+    });
```
NIT: might be interesting to mix nested arrays and objects along with aliases, just for completeness.

crisbeto replied:
Added one more test case that's a bit more convuluted.

JoostK commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts:
```diff
@@ -598,6 +599,47 @@ export class StaticInterpreter {
+          if (ts.isIdentifier(name)) {
+            key = name.text;
+          } else {
+            return DynamicValue.fromUnknown(element);
```
Dynamic values typically capture the full chain of walked nodes, such that the remaining tail of `path` would also be captured. (also the case for the dynamic value below)

crisbeto replied:
I'm not sure I follow what you mean here.

JoostK replied:
As the static interpreter walks the AST nodes, it may introduce a `DynamicValue` for a node that it can't evaluate statically. Every node leading up to that specific node will wrap that dynamic value using `DynamicValue.forDynamicInput`, building a trail of nodes that has been walked. This information can be used in diagnostics to show the path that the compiler has taken to better explain what the problem is (although this is still pending work atm, I recently started working on it).

If you consider `const {a: { body }}` = { a: window };` and trying to evaluate `body` this wil come into play: the outermost `DynamicValue` would arrive at `{ body }`, which wraps the `DynamicValue` for `a` which points at `window`.

After some more thought I think we should actually keep it simple for now, so we keep this as is. If we do see room for improvement we can always still reconsider this case.

crisbeto replied:
Great, and thank you for the explanation 👍

JoostK commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts:
```diff
@@ -598,6 +599,47 @@ export class StaticInterpreter {
+          }
+        }
+        value = this.accessHelper(element, value, key, context);
+        if (value instanceof DynamicValue) {
+          return value;
```
This one should be fairly simple to add a test for, e.g. using `const {a} = {a: window};`. If you were to extend that to `const {a: {b}} = {a: {b: window}};` you'll see how chaining the dynamic values becomes relevant.

crisbeto replied:
Added.

crisbeto commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts:
```diff
@@ -598,6 +599,47 @@ export class StaticInterpreter {
+      closestDeclaration = closestDeclaration.parent;
+    }
+
+    if (ts.isVariableDeclaration(closestDeclaration) &&
+        closestDeclaration.initializer !== undefined) {
```
The condition can't be inverted, because the `isVariableDeclaration` is what casts it to "something that has an `initializer`".

JoostK replied:
Would this not work?

```suggestion
    if (!ts.isVariableDeclaration(closestDeclaration) ||
        closestDeclaration.initializer === undefined) {
      return DynamicValue.fromUnknown(node);
    }
```

crisbeto replied:
Ah ok, I see. I understood the "early return" above as short-circuiting the `if` statement.

crisbeto committed:
fix(compiler): unable to resolve destructuring variable declarations

Currently the partial evaluator isn't able to resolve a variable declaration that uses destructuring in the form of `const {value} = {value: 0}; const foo = value;`. These changes add some logic to allow for us to resolve the variable's value.

Fixes #36917.

JoostK reviewed (approved):
Nicely done! 🎉
